posts/python-in-production-part-5-of-5.md

```diff
@@ -0,0 +1,250 @@
+
+Choose Git under SCM, fill out your Repository URL, skip the Credential section, because this is a public repo, and then specify the branch to build.  Branch is typically `*/master` or `*/main`.   Leave the rest of the options as they are and press **save**.
+
+# Run the Build Job
+
+Now that you have created your Jenkins job, which is a pipeline job, we are ready to run it!  Navigate back to the main dashboard and click on your Jenkins Job.  On the Job screen, you will see some options on the left.  Click `Build Now`.  Further down on the left side bar, you should see your job pop up in the queue.  If you click on the job, you can navigate to the `Console Output` screen.  This will show you the `stdout` of your build job.
+
+If the job passes, you will see `Finished: SUCCESS` at the bottom of the console output.  Chances are, you may need to iterate over you build script and Jenkins setup a few times before you get a successful build.  Let's look at some troubleshooting steps.
+
+## Build Environment
+
+Jenkins runs in a linux vm.  It can be ran directly on a machine or inside a dockerfile.  Because of this reason, it is constrained by the machine it is running on.  During the build process, you may have gotten an error that Python3 was not installed.  In order to get around this, we need to add python to the build machine.  
+
+If you are using Docker, then update your dockerfile to include an installation step, such as this.
+```dockerfile
+FROM jenkins/jenkins:lts-jdk11
+USER root
+RUN apt-get update && apt-get install -y python3 && apt-get install -y python3-venv
+USER jenkins
+```
+
+This Dockerfile adds a few needed dependencies to the docker image.  Running `docker build` on this file and then using it as your Jenkins instance should get rid of any dependency issues you have.  **Note:** I had to downgrade my sample-python-project to python 3.9 in order to build this on my debian machine, as Python 3.10 is a bit of a pain to install.  
+
+If you are using Jenkins directly on a virtual machine, you can simply ssh into that machine and install what ever needed dependencies you have.
+
+# Download the Built Binary
+
+Once your Jenkins job has completed, you may be wondering, where is my Binary?  The simple way to get your file off the Jenkins Build machine and onto your local machine, is by using the Jenkins UI.  
+
+If you are still on the console output screen, then you are already half way there.  Navigate to the `Workspaces` folder on the left side bar.  You should see a workspace directory in the center of the screen.  If you click on the path, you should now see a familiar site, our project directory.  If you simply click on `simple_api`, it will initiate a file download in your browser.  There are *better* ways to make this happen however.
```

Comment:
```suggestion
You are already halfway there if you are still on the console output screen.  Navigate to the `Workspaces` folder on the left sidebar.  You should see a workspace directory in the center of the screen.  If you click on the path, you should now see a familiar site, our project directory.  If you click on `simple_api`, it will initiate a file download in your browser.  There are *better* ways to make this happen, however.
```